### PR TITLE
SI-8806 Add lower bound check to Any lint

### PR DESCRIPTION
**Pull request by som-snytt**
We already exclude the lint check for infer-any if
Any is somewhere explicit.

This commit adds lower bounds of type params to
the somewheres.

Apologies for the "royal we" in the commit message.

Review by @retronym

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 1]
```diff
@@ -546,7 +546,14 @@ trait Infer extends Checkable {
       val targs = solvedTypes(tvars, tparams, tparams map varianceInTypes(formals), upper = false, lubDepth(formals) max lubDepth(argtpes))
       // Can warn about inferring Any/AnyVal as long as they don't appear
       // explicitly anywhere amongst the formal, argument, result, or expected type.
-      def canWarnAboutAny = !(pt :: restpe :: formals ::: argtpes exists (t => (t contains AnyClass) || (t contains AnyValClass)))
+      // ...or lower bound of a type param, since they're asking for it.
+      def canWarnAboutAny = {
+        val loBounds = tparams map (_.typeSignature) collect { case TypeBounds(lo, _) => lo }
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: Let's stick to `.info` in the compiler. We can also use the convenience method `.bounds`.

``` scala
scala> class C[A >: Any]
defined class C

scala> typeOf[C[_]].typeSymbol.typeParams.map(_.info.bounds.lo)
res10: List[$r.intp.global.Type] = List(Any)
```

Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: Thanks, it sure is nice knowing a guy on the inside.

**Comment by gkossakowski** [comment]
Ping @retronym.

**Comment by som-snytt** [comment]
@retronym, this is to avoid:

```
scala> f"${42}"
res0: String = 42

scala> :replay -Xlint
Replaying: f"${42}"
<console>:8: warning: a type was inferred to be `Any`; this may indicate a programming error.
              f"${42}"
                  ^
res0: String = 42

```

**Comment by retronym** [comment]
LGTM !

**Pull request merged**